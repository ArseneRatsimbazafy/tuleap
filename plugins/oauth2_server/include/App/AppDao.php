<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Tuleap\DB\DataAccessObject;

class AppDao extends DataAccessObject
{
    /**
     * @psalm-return array{id:int, project_id:int, name:string, redirect_endpoint: string}
     */
    public function searchByClientId(ClientIdentifier $client_id): ?array
    {
        $sql = 'SELECT * FROM plugin_oauth2_server_app
            WHERE id = ?';
        return $this->getDB()->row($sql, $client_id->getInternalId());
    }

    /**
     * @psalm-return array<array{id:int, project_id:int, name:string, redirect_endpoint: string}>
     */
    public function searchByProject(\Project $project): array
    {
        $sql = "SELECT * FROM plugin_oauth2_server_app
            WHERE project_id = ?";
        return $this->getDB()->run($sql, $project->getID());
    }

    public function create(NewOAuth2App $app): void
    {
        $this->getDB()->insert(
            'plugin_oauth2_server_app',
            [
                'project_id'        => $app->getProject()->getID(),
                'name'              => $app->getName(),
                'redirect_endpoint' => $app->getRedirectEndpoint()
            ]
        );
    }

    public function delete(int $app_id): void
    {
        $this->getDB()->run('DELETE FROM plugin_oauth2_server_app WHERE id = ?', $app_id);
    }
}
