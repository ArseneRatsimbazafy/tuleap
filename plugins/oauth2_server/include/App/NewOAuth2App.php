<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Tuleap\Authentication\SplitToken\SplitTokenVerificationString;
use Tuleap\Authentication\SplitToken\SplitTokenVerificationStringHasher;

/**
 * @psalm-immutable
 */
final class NewOAuth2App
{
    /**
     * @var string
     */
    private $name;
    /**
     * @var string
     */
    private $redirect_endpoint;
    /**
     * @var SplitTokenVerificationString
     */
    private $secret;
    /**
     * @var string
     */
    private $hashed_secret;
    /**
     * @var \Project
     */
    private $project;

    private function __construct(
        string $name,
        string $redirect_endpoint,
        SplitTokenVerificationString $secret,
        string $hashed_secret,
        \Project $project
    ) {
        $this->name              = $name;
        $this->redirect_endpoint = $redirect_endpoint;
        $this->secret            = $secret;
        $this->hashed_secret     = $hashed_secret;
        $this->project           = $project;
    }

    /**
     * @throws InvalidAppDataException
     */
    public static function fromAppData(
        string $name,
        string $redirect_endpoint,
        \Project $project,
        SplitTokenVerificationStringHasher $hasher
    ): self {
        $is_data_valid = self::isAppDataValid($name, $redirect_endpoint);

        if (! $is_data_valid) {
            throw new InvalidAppDataException();
        }

        $secret = SplitTokenVerificationString::generateNewSplitTokenVerificationString();

        return new self(
            $name,
            $redirect_endpoint,
            $secret,
            $hasher->computeHash($secret),
            $project
        );
    }

    private static function isAppDataValid(string $name, string $redirect_endpoint): bool
    {
        $string_validator = new \Valid_String();
        $string_validator->required();
        // See https://tools.ietf.org/html/rfc6749#section-3.1.2
        $redirect_endpoint_validator = new \Valid_String();
        $redirect_endpoint_validator->required();
        $redirect_endpoint_validator->addRule(new \Rule_Regexp('/^https:\/\/[^#]*$/i'));

        return $string_validator->validate($name) && $redirect_endpoint_validator->validate($redirect_endpoint);
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getRedirectEndpoint(): string
    {
        return $this->redirect_endpoint;
    }

    public function getProject(): \Project
    {
        return $this->project;
    }

    public function getSecret(): SplitTokenVerificationString
    {
        return $this->secret;
    }

    public function getHashedSecret(): string
    {
        return $this->hashed_secret;
    }
}
