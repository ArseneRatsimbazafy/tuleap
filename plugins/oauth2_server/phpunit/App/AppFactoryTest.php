<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Mockery as M;
use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
use PHPUnit\Framework\TestCase;

final class AppFactoryTest extends TestCase
{
    use MockeryPHPUnitIntegration;

    /**
     * @var AppFactory
     */
    private $app_factory;
    /**
     * @var M\LegacyMockInterface|M\MockInterface|AppDao
     */
    private $app_dao;

    protected function setUp(): void
    {
        $this->app_dao     = M::mock(AppDao::class);
        $this->app_factory = new AppFactory($this->app_dao);
    }

    public function testGetAppsForProject(): void
    {
        $rows    = [
            ['id' => 1, 'name' => 'Jenkins'],
            ['id' => 2, 'name' => 'My custom REST client']
        ];
        $project = M::mock(\Project::class);
        $this->app_dao->shouldReceive('searchByProject')
            ->once()
            ->with($project)
            ->andReturn($rows);

        $result = $this->app_factory->getAppsForProject($project);
        $this->assertEquals(
            [new OAuth2App(1, 'Jenkins', $project), new OAuth2App(2, 'My custom REST client', $project)],
            $result
        );
    }
}
