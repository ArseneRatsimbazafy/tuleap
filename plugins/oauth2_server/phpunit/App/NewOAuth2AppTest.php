<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Mockery as M;
use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
use PHPUnit\Framework\TestCase;

final class NewOAuth2AppTest extends TestCase
{
    use MockeryPHPUnitIntegration;

    public function testFromAppDataThrowsWhenAppNameIsEmpty(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('', 'https://example.com/redirect', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointIsEmpty(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', '', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointIsNotHTTPS(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', 'http://insecure.example.com', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointContainsAnAnchor(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', 'https://example.com/redirect#fragment', M::mock(\Project::class));
    }

    public function testFromAppDataReturnsANewOauth2AppToBeSavedInDatabase(): void
    {
        $app_name          = 'Jenkins';
        $redirect_endpoint = 'https://example.com/redirect';
        $project           = M::mock(\Project::class);
        $new_app           = NewOAuth2App::fromAppData($app_name, $redirect_endpoint, $project);
        $this->assertSame($app_name, $new_app->getName());
        $this->assertSame($redirect_endpoint, $new_app->getRedirectEndpoint());
        $this->assertSame($project, $new_app->getProject());
    }

    public function testFromAppDataAllowsRedirectEndpointWithQuery(): void
    {
        $redirect_endpoint = 'https://example.com/redirect?key=value';
        $new_app = NewOAuth2App::fromAppData('Jenkins', $redirect_endpoint, M::mock(\Project::class));
        $this->assertSame($redirect_endpoint, $new_app->getRedirectEndpoint());
    }
}
