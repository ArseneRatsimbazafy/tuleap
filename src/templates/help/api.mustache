<h1>Tuleap API</h1>

<section class="tlp-pane">
    <div class="tlp-pane-container">
        <div class="tlp-pane-header">
            <h1 class="tlp-pane-title">REST API</h1>
        </div>
        <section class="tlp-pane-section">
            <p>The recommended way to interact with Tuleap is to leverage on the REST
                API. REST API has broad test coverage in Tuleap developers test suite and
                there is a large number of end points. Modern features (Kanban, Scrum Planning,
                etc) are using those APIs.</p>

            <p>You can <a href="/doc/en/user-guide/integration/rest.html">get started</a> from the
                documentation.<p>

            {{# explorer_url }}
                <p>You can also <a href="{{ explorer_url }}">discover and test online</a> the
                    various routes.</p>
            {{/ explorer_url }}
        </section>
    </div>
</section>

{{# should_display_documentation_about_deprecated_soap_api }}
<section class="tlp-pane">
    <div class="tlp-pane-container">
        <div class="tlp-pane-header">
            <h1 class="tlp-pane-title">Legacy APIs</h1>
        </div>
        <section class="tlp-pane-section">
            <div class="tlp-alert-warning">Those APIs are deprecated. Whenever possible
                there shouldn't be any new developments done relying on them.</div>

            <h2 class="tlp-pane-subtitle">SOAP</h2>

            <p>More informations available in the <a href="https://tuleap.net/wiki/index.php?group_id=101&pagename=HowToUseTuleapSoapApi">documentation</a> on Tuleap.net.</p>

            <ul>
            {{# end_points }}
                <li>{{ title }} - v{{ version }}
                    <p>{{{ description }}}</p>
                    <ul>
                        <li><a href="{{ changelog }}">ChangeLog</a></li>
                        <li><a href="{{ wsdl_viewer }}">WSDL Documentation</a></li>
                        <li><a href="{{ wsdl }}">WSDL</a></li>
                    </ul>
                </li>
            {{/ end_points }}
            </ul>
        </section>
    </div>
</section>
{{/ should_display_documentation_about_deprecated_soap_api }}
