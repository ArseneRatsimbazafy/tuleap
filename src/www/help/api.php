<?php

require_once 'pre.php';

$renderer = TemplateRendererFactory::build()->getRenderer(ForgeConfig::get('codendi_dir') . '/src/templates/help');



$presenter = array(
    'should_display_documentation_about_deprecated_soap_api' => ForgeConfig::get('should_display_documentation_about_deprecated_soap_api'),
    'explorer_available' => is_dir('/usr/share/tuleap/src/www/api/explorer'),
    'end_points'         => array(
        array(
            'title'       => 'Core',
            'wsdl'        => '/soap/?wsdl',
            'wsdl_viewer' => '/soap/wsdl.php',
            'changelog'   => '/soap/ChangeLog',
            'version'     => file_get_contents(dirname(__FILE__).'/../soap/VERSION'),
            'description' => <<<EOT
Historically the sole end point, therefore it groups multiple different functions:
<ul>
    <li>Session management: login, logout, projects, ...</li>
    <li>File Release System access (FRS): addPackage, addRelease, addFile, ...</li>
    <li>Tracker v3 (for historical deployments): get/updateTracker, get/updateArtifact, ...</li>
    <li>Documentation: get/updateDocman, ...</li>
</ul>
EOT
        ),
        array(
            'title'       => 'Subversion',
            'wsdl'        => '/soap/svn/?wsdl',
            'wsdl_viewer' => '/soap/svn/wsdl-viewer.php',
            'changelog'   => '/soap/svn/ChangeLog',
            'version'     => file_get_contents(dirname(__FILE__).'/../soap/svn/VERSION'),
            'description' => 'Get informations about Subversion usage in project.',
        ),
        array(
            'title'       => 'Project',
            'wsdl'        => '/soap/project/?wsdl',
            'wsdl_viewer' => '/soap/project/wsdl-viewer.php',
            'changelog'   => '/soap/project/ChangeLog',
            'version'     => file_get_contents(dirname(__FILE__).'/../soap/project/VERSION'),
            'description' => 'Create and administrate projects.',
        ),
    )
);

$GLOBALS['HTML']->header(array('title' => 'API', 'main_classes' => array('tlp-framed')));
$renderer->renderToPage('api', $presenter);
$GLOBALS['HTML']->footer(array());
