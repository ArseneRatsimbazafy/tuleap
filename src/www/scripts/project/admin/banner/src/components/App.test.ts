/*
 * Copyright (c) Enalean, 2019-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import App from "./App.vue";
import BannerDeleter from "./BannerDeleter.vue";
import { createProjectAdminBannerLocalVue } from "../helpers/local-vue-for-tests";
import * as rest_querier from "../api/rest-querier";

describe("App", () => {
    it("displays something when no banner is set", async () => {
        const wrapper = shallowMount(App, {
            localVue: await createProjectAdminBannerLocalVue(),
            propsData: {
                message: "",
                project_id: 108
            }
        });

        expect(wrapper.text()).not.toBe("");
    });

    it("displays message and remove button when banner is not empty", async () => {
        const banner_message = "<b>My banner content</b>";

        const wrapper = shallowMount(App, {
            localVue: await createProjectAdminBannerLocalVue(),
            propsData: {
                message: banner_message,
                project_id: 108
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("Should be able to send the deletion request", async () => {
        const wrapper = shallowMount(App, {
            localVue: await createProjectAdminBannerLocalVue(),
            propsData: {
                message: "some message",
                project_id: 108
            }
        });

        const reload_page = jest.spyOn(window.location, "reload").mockImplementation();
        const delete_banner = jest
            .spyOn(rest_querier, "deleteBannerForProject")
            .mockImplementation(() => {
                return Promise.resolve();
            });

        wrapper.find(BannerDeleter).vm.$emit("delete-banner");

        await wrapper.vm.$nextTick();

        expect(delete_banner).toHaveBeenCalledTimes(1);
        expect(reload_page).toHaveBeenCalledTimes(1);
    });

    it("Should display an error if banner deletion fails", async () => {
        const wrapper = shallowMount(App, {
            localVue: await createProjectAdminBannerLocalVue(),
            propsData: {
                message: "some message",
                project_id: 108
            }
        });

        jest.spyOn(rest_querier, "deleteBannerForProject").mockImplementation(() => {
            return Promise.reject(new Error("an error message"));
        });

        wrapper.find(BannerDeleter).vm.$emit("delete-banner");

        await wrapper.vm.$nextTick();

        expect(wrapper.element).toMatchSnapshot();
    });
});
