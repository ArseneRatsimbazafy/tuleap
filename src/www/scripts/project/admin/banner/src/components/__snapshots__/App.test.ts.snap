// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should display an error if banner deletion fails 1`] = `
<div>
  <div
    class="tlp-alert-danger"
    data-current-language="en_US"
    data-msgid="
        An error occurred: %{ error_message }
    "
  >
    
        An error occurred: an error message
    
  </div>
   
  <div>
    <banner-presenter-stub
      message="some message"
    />
     
    <banner-deleter-stub />
  </div>
</div>
`;

exports[`App displays message and remove button when banner is not empty 1`] = `
<div>
  <!---->
   
  <div>
    <banner-presenter-stub
      message=<b>My banner content</b>
    />
     
    <banner-deleter-stub />
  </div>
</div>
`;
